Automattic/jetpack
# Jetpack AI: Support tiered plan usage on the Usage Panel

Fixes #33887.
Fixes #34124.
Fixes #33845.

## Proposed changes:

* Change the usage control component to support plans with tiers
   * For plans with tiers, the usage info comes from the current period usage, the requests limit comes from the current tier and the usage control needs to display a reset date for the requests (a counter)
* Fix design issues when handling the free and and the unlimited plans (pe4Cmx-1W0-p2#comment-1015)

**Note:** the Figma file shows a different text for the upgrade button for tiered plans. This will be handled on a follow-up.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Let's test all the variations by dispatching actions that will update the AI Assistant feature data on the store:
* Apply this branch to your test site, go to the editor and open the usage panel on the Jetpack sidebar > AI Assistant:

<img width="300" alt="Screenshot 2023-11-17 at 18 48 02" src="https://github.com/Automattic/jetpack/assets/6760046/d5e02abd-a762-4e93-a3d3-74fcbb1adab8">

* Open the browser's console so we can dispatch testing actions
* For every test case below, check if the UI is following the design proposed for tiers on the Figma file ([link]) or the fixed versions for the free and unlimited plans (Jq2UFBOjX0U5fYmAoCrwaV-fi-4076_7023); an image of the expected behavior is showed with each test case

### Free plan below the limit

```
wp.data.dispatch( 'wordpress-com/plans' ).storeAiAssistantFeature( {
    "hasFeature": false,
    "isOverLimit": false,
    "requestsCount": 15,
    "requestsLimit": 20,
    "requireUpgrade": false,
    "upgradeType": "default",
    "usagePeriod": {
        "currentStart": null,
        "nextStart": null,
        "requestsCount": 15
    },
    "currentTier": {
        "value": 0,
	"limit": 20,
    },
    "nextTier": {
        "slug": "ai-assistant-tier-100",
        "limit": 100,
        "value": 100
    }
} );
```

<img width="250" alt="Screenshot 2023-11-17 at 19 06 16" src="https://github.com/Automattic/jetpack/assets/6760046/f157d661-9c83-4194-a609-4b241341099b">

### Free plan on the limit

```
wp.data.dispatch( 'wordpress-com/plans' ).storeAiAssistantFeature( {
    "hasFeature": false,
    "isOverLimit": true,
    "requestsCount": 20,
    "requestsLimit": 20,
    "requireUpgrade": true,
    "upgradeType": "default",
    "usagePeriod": {
        "currentStart": null,
        "nextStart": null,
        "requestsCount": 20
    },
    "currentTier": {
        "value": 0,
	"limit": 20,
    },
    "nextTier": {
        "slug": "ai-assistant-tier-100",
        "limit": 100,
        "value": 100
    }
} );
```

<img width="250" alt="Screenshot 2023-11-17 at 19 07 27" src="https://github.com/Automattic/jetpack/assets/6760046/f266cbbe-9fa2-448b-9df1-64c1e9e9a9c6">

### Unlimited plan

```
wp.data.dispatch( 'wordpress-com/plans' ).storeAiAssistantFeature( {
    "hasFeature": true,
    "isOverLimit": false,
    "requestsCount": 15,
    "requestsLimit": 20,
    "requireUpgrade": false,
    "upgradeType": "default",
    "usagePeriod": {
        "currentStart": "2023-11-06 00:00:00",
        "nextStart": "2023-12-06 00:00:00",
        "requestsCount": 15
    },
    "currentTier": {
        "value": 1,
	"limit": 999999999,
    },
    "nextTier": {
        "slug": "ai-assistant-tier-100",
        "limit": 100,
        "value": 100
    }
} );
```

<img width="250" alt="Screenshot 2023-11-17 at 18 59 23" src="https://github.com/Automattic/jetpack/assets/6760046/7426e247-52d2-4ab5-a323-1dbee3c8f90e">

### Tiered plan below the limit

```
wp.data.dispatch( 'wordpress-com/plans' ).storeAiAssistantFeature( {
    "hasFeature": true,
    "isOverLimit": false,
    "requestsCount": 345,
    "requestsLimit": 20,
    "requireUpgrade": false,
    "upgradeType": "default",
    "usagePeriod": {
        "currentStart": "2023-11-06 00:00:00",
        "nextStart": "2023-12-06 00:00:00",
        "requestsCount": 80
    },
    "currentTier": {
        "slug": "ai-assistant-tier-100",
        "limit": 100,
        "value": 100
    },
    "nextTier": {
        "slug": "ai-assistant-tier-200",
        "limit": 200,
        "value": 200
    }
} );
```

<img width="250" alt="Screenshot 2023-11-20 at 17 47 55" src="https://github.com/Automattic/jetpack/assets/6760046/cdee8564-d5a9-4812-88a8-2108f2df6ab7">

**Note 1:** confirm the reset counter is correct. It counts the amount of days from today until the `usagePeriod.nextStart` date.

**Note 2:** the Figma file shows a different text for the upgrade button in this scenario. This will be handled on a follow-up.

### Tiered plan on the limit

```
wp.data.dispatch( 'wordpress-com/plans' ).storeAiAssistantFeature( {
    "hasFeature": true,
    "isOverLimit": true,
    "requestsCount": 345,
    "requestsLimit": 20,
    "requireUpgrade": true,
    "upgradeType": "default",
    "usagePeriod": {
        "currentStart": "2023-11-06 00:00:00",
        "nextStart": "2023-12-06 00:00:00",
        "requestsCount": 100
    },
    "currentTier": {
        "slug": "ai-assistant-tier-100",
        "limit": 100,
        "value": 100
    },
    "nextTier": {
        "slug": "ai-assistant-tier-200",
        "limit": 200,
        "value": 200
    }
} );
```

<img width="250" alt="Screenshot 2023-11-20 at 17 48 20" src="https://github.com/Automattic/jetpack/assets/6760046/f82e5e69-7bf7-46f6-92a2-33db6307c8c5">

**Note 1:** confirm the reset counter is correct. It counts the amount of days from today until the `usagePeriod.nextStart` date.

**Note 2:** the Figma file shows a different text for the upgrade button in this scenario. This will be handled on a follow-up.

---

You can test some other combination of status by dispatching the respective action. You can also check if the usage panel is showing the correct UI for the current status of your test sites. An extra step is to hardcode usage values on the backend code (sandboxing your API) and then checking the results on the site.
